### PR TITLE
275 feature implementation of coral

### LINKED ISSUE #275: feature: Implementation of CORAL
**Purpose**
Implementation of CORAL (CORrelation ALignment) [1], a domain adaptation 
method that aligns the second-order statistics (covariance) of source and 
target domains.

**Brief description**
CORAL minimizes domain shift by whitening the target data and re-coloring 
it with the source covariance structure, enabling calibration models trained 
on source data to remain valid on target instruments.

**Proposed functionality**
- Alignment of source and target covariance matrices
- Support for regularization parameter to handle singular covariance matrices
- Consistent API for fit / transform

**Reference**
[1] Sun, B., Feng, J., & Saenko, K. (2016).
Return of Frustratingly Easy Domain Adaptation.
Proceedings of the AAAI Conference on Artificial Intelligence, 30(1).
https://arxiv.org/abs/1511.05547

### PR DESCRIPTION
**Pull request by inarteroger**
## Why is this change needed?

Adds the `CORAL` (CORrelation ALignment) and `SubspaceAlignment` (SA) transformers to the `chemotools.adaptation` module. Both are unsupervised linear domain adaptation methods that align target spectral data to a source (reference) domain without requiring labelled target samples. This extends the adaptation module alongside the existing DS, PDS and SST transformers.

## Type of change

* New feature

## Description

Implementation of `CORAL` following the method described in:

Sun, B., Feng, J., & Saenko, K. (2016). Return of Frustratingly Easy Domain Adaptation. *Proceedings of the AAAI Conference on Artificial Intelligence*, 30(1). https://doi.org/10.1609/aaai.v30i1.10306

Implementation of `SubspaceAlignment` following the method described in:

Fernando, B., Habrard, A., Sebban, M., & Tuytelaars, T. (2013). Unsupervised Visual Domain Adaptation Using Subspace Alignment. *IEEE International Conference on Computer Vision (ICCV)*. https://doi.org/10.1109/ICCV.2013.368

## What was added

* `CORAL` transformer in `chemotools/adaptation/_coral.py`
* `SubspaceAlignment` transformer in `chemotools/adaptation/_subspace_alignment.py`
* Full sklearn API compliance for both transformers (`fit`, `transform`, `fit_transform`)
* Identity fallback when `X_source=None`

## Tests added

* sklearn estimator compliance
* fit attributes and shape checks
* boundary and edge case validation
* Pipeline + GridSearchCV + metadata routing integration

### CODE REVIEW

File: chemotools/adaptation/_subspace_alignment.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
+"""
+The :mod: `chemotools.adaptation._subspace_alignment`
+module implements the Subspace aslignment (SA) transformer
```
[the remark this answers is not in the document]
Comment (inarteroger) [reply chemotools/adaptation/_subspace_alignment.py]: Done

File: chemotools/adaptation/_subspace_alignment.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+from chemotools._doc_mixin import DocLinkMixin
+
+
+class SubspaceAlignment(
+    DocLinkMixin, OneToOneFeatureMixin, TransformerMixin, BaseEstimator
+):
+    """
+    Subspace Alignment (SA) is an unsupervised domain adaptation transformer
+    hat aligns the principal subspaces of source and target feature
```
[the remark this answers is not in the document]
Comment (inarteroger) [reply chemotools/adaptation/_subspace_alignment.py]: Done

File: chemotools/adaptation/_coral.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            self.C_X_source_sqrt_ = None
+            self.x_source_provided_ = False
+            return self
+
+        # Validate X_source as a plain array (do not update n_features_in_)
+        X_source = np.asarray(X_source, dtype=np.float64)
+        if X_source.ndim != 2:
+            raise ValueError(
+                f"X_source must be a 2D array, got {X_source.ndim}D array."
+            )
```
[the remark this answers is not in the document]
Comment (inarteroger) [reply chemotools/adaptation/_coral.py]: Changed 
```python
X_source = np.asarray(X_source, dtype=np.float64)
```
to
```python
X_source = validate_data(
             self, X_source, ensure_2d=True, reset=False, dtype=np.float64
         )
```

File: chemotools/adaptation/_coral.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        self.X_source_mean_ = X_source.mean(axis=0)
+        self.X_source_centered_ = X_source - self.X_source_mean_
+
+        # Covariance matrix
+        self.C_X_ = np.cov(self.X_centered_, rowvar=False, ddof=1) + self.reg * np.eye(
+            self.n_features_in_
+        )
+        self.C_X_source_ = np.cov(
+            self.X_source_centered_, rowvar=False, ddof=1
+        ) + self.reg * np.eye(self.n_features_in_)
```
[the remark this answers is not in the document]
Comment (inarteroger) [reply chemotools/adaptation/_coral.py:213]: Add after `validate_data`:
```python
if X.shape[0] < 2:
    raise ValueError(
        f"X must have at least 2 samples, got {X.shape[0]}."
    )
if X_source.shape[0] < 2:
    raise ValueError(
        f"X_source must have at least 2 samples, got {X_source.shape[0]}."
    )
```
and the relative test:
```python
class TestFit:
    """Test for the fit method behavior."""
    def test_coral_raises_if_single_sample(self):
        X = np.array([[1.0, 2.0]])
        X_source = np.array([[3.0, 4.0]])
        with pytest.raises(ValueError, match="at least 2 samples"):
            CORAL().fit(X, X_source=X_source)
```

File: chemotools/adaptation/_subspace_alignment.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        # Check consistency between X and X_source
+        # if X_source.shape != X.shape:
+        #     raise ValueError(
+        #         f"X and X_source must have the same shape, "
+        #         f"got X={X.shape} and X_source={X_source.shape}."
+        #     )
+        self.X_mean_ = X.mean(axis=0)
+        self.X_std_ = X.std(axis=0)
+        self.X_source_mean_ = X_source.mean(axis=0)
+        self.X_source_std_ = X_source.std(axis=0)
```
[the remark this answers is not in the document]
Comment (inarteroger) [reply chemotools/adaptation/_subspace_alignment.py]: Changed
```python
self.X_std_ = X.std(axis=0)
self.X_source_std_ = X_source.std(axis=0)
```
to
```python
self.X_std_ = np.where(X.std(axis=0) == 0, 1.0, X.std(axis=0))
self.X_source_std_ = np.where(
    X_source.std(axis=0) == 0, 1.0, X_source.std(axis=0)
)
```

**Comment by inarteroger** [comment]
Addressed all Copilot suggestions
